### PR TITLE
feat: add support for model from other libraries

### LINKED ISSUE #1786: Add more tests cases for scarb projects
**Is your feature request related to a problem? Please describe.**
The spawn-and-move example is currently too simple. And we miss some use cases like:
1. Compiling a library
2. Compiling a workspace
3. Contracts with same name but not the same path
4. Using external contracts

**Describe the solution you'd like**
We should add more projects into example to break down the cases we're not covering yet. This will help detecting errors during the tests earlier, and some edge cases of Scarb compiler.

**Additional context**
We should add different project to cover specific things, to ensure things can be test easily and separately.

### LINKED ISSUE #2111: Add contract registry to the world
#1956 was a first good work to get the contract registry. This should be integrated in the namespace work, which will ease this integration.

Basically, we need to enable the world to keep track of the contracts that are registered, to be callable from other contracts without knowing the address in advance.

Doing so, one can easily access a contract's address form the world without hard-coding the address.
We could also add a layer of permission here, with the `writer` permission, or adding an `executor` permission, to kind of control which contract can call which contract, which unlocks some new horizons.

### LINKED ISSUE #2151: Debug the library compilation with namespaces
**Is your feature request related to a problem? Please describe.**
With the `1.0.0-alpha.0`, if a library like Origami is built imported in a project (with `build-external-contract`), the models/contracts are not correctly generated.

**Describe the solution you'd like**
We should be able to build a library, and also verify if the namespace applied is the one from the library itself, or the project that imports it. We **must** put the namespace of the current project to override the one of the library.

### PR DESCRIPTION
**Pull request by glihm**
# Description

This PR enables Dojo models/contracts from an other projects to be imported with current project namespace. This opens door to more reusability of models, like Origami.

The current limitation is that, when writing a library of Dojo models/contracts, the `namespace` argument of the `dojo::model` or `dojo::contract` must be specified. This is because inside `dojo-lang`, when the model/contract is processed by the Dojo plugin, we don't have any relationship between the code and the crates from which it comes from.

```rust
#[dojo::model(namespace: "mylib")]
struct MyLibModel { ... }
```
Doing so, anytime a model/contract is processed by the Dojo plugin, we can know the origin of the model/contract, and apply some mappings.

To use a model from a library, the steps are:
1. The `build-external-contracts` must be used, with the exact same casing that was defined, and the fully qualified path of the model/contract. We force the snake case under the hood to follow the compiler nomenclature, but the user can use the exact same casing.
2. The namespace is now a struct `NamespaceConfig`, where one must define the default namespace and some mappings can be added. If no mapping is used, the default namespace of the crate root is used. For the namespace mapping, we use the model/contract tag.

## Example

```toml
# Scarb.toml
[dependencies]
armory = { path = "../game-lib/armory" }
```

```rust
// ./game-lib/armory/src/lib.cairo
#[dojo::model(namespace: "armory")]
struct FlatBow {...}
```
```toml
# ./spawn-and-move/Scarb.toml
build-external-contracts = [
    "armory::FlatBow",
]

[tool.dojo.world.namespace]
default = "dojo_examples"
# This mapping is only valid for the specified tag.
# The namespace for `FlatBow` will be `dojo_examples_weapons`.
mappings = { "armory-FlatBow" = "dojo_examples_weapons" }
```
If the whole models of a library may go in the same namespace, then only the namespace
can be specified in the mapping:
```toml
[tool.dojo.world.namespace]
default = "dojo_examples"
# All the models/contracts inside `armory` will use `dojo_examples_weapons` as namespace.
mappings = { "armory" = "dojo_examples_weapons" }
```
Finally, anything that is not found into the mappings, will use the default namespace.
This means, any library model without a specified namespace in the attribute will use the default namespace of the crate root.

## Related issue

Closes #1786.
Closes #2151.
Closes #2111.

## Tests

Compiler utility function has been reworked to support multi package copying to the tmp folder. This also support namespace to enable compilation of workspaces in tests.

## Summary by CodeRabbit

- **New Features**
  - Introduced a new inline macro module for generating selectors based on tags.

- **Refactor**
  - Replaced `package_id` with `namespace_config` across various modules to streamline namespace handling.

- **Bug Fixes**
  - Fixed issues related to attribute errors and unsupported attributes in system tests.

- **Chores**
  - Updated `dojo-world.json` and `dojo-world.toml` files to improve function clarity and update class hashes.

- **Tests**
  - Added error messages for missing namespaces and plugins in test data.

### CODE REVIEW

File: bin/sozo/src/commands/migrate.rs [hunk 1]
```diff
@@ -88,36 +88,42 @@ impl MigrateArgs {
...
+                        rpc_url,
+                        account,
+                        &name,
+                        true,
+                        TxnConfig::default(),
+                        dojo_metadata.skip_migration,
+                    )
+                    .await
+                })
+                .map(|_| ()),
```
[the remark this answers is not in the document]
Comment (glihm) [reply bin/sozo/src/commands/migrate.rs:107]: The outer function is expecting a result with `()` unit type, but in our case, we now return `MigrationOutput`, which in other places can be very useful to access.

Using `.map(|_| ())` ensures that whatever is returned with `Ok`, we return the unit type discarding the returned value.

File: examples/spawn-and-move/src/actions.cairo [hunk 2]
```diff
@@ -30,8 +28,15 @@ mod actions {
...
-    use dojo_examples::something::{ISomethingDispatcher, ISomethingDispatcherTrait};
+
+    // Featues can be used on modules, structs, trait and `use`. Not inside
+    // a function.
+    #[cfg(feature: 'dungeon')]
+    use dojo_examples::dungeon::{IDungeonDispatcher, IDungeonDispatcherTrait};
+    #[cfg(feature: 'dungeon')]
+    use armory::Flatbow;
+    #[cfg(feature: 'dungeon')]
+    use bestiary::RiverSkale;
```
[the remark this answers is not in the document]
Comment (glihm) [reply examples/spawn-and-move/src/actions.cairo:39]: @remybar made a comment and we will update to ensure any library imported can be placed under a different namespace if the user wants to.

So currently what happens, is that the model is imported, and at the compilation time, the namespace of the current package is used instead of the default namespace of the library.

For this reason, if someone is forcing the namespace like `#[dojo::model(namespace: "nm1")]`, hence the namespace of the model **even if imported elsewhere** will be `nm1`. This is due to how we define the namespace during the plugin expansion phase.

File: crates/dojo-test-utils/src/compiler.rs [hunk 3]
```diff
@@ -149,7 +202,11 @@ pub fn copy_project_temp(
 
                 let mut table = contents.parse::<Table>().expect("Failed to parse Scab.toml");
 
-                let dojo = table["dependencies"]["dojo"].as_table_mut().unwrap();
+                let dojo = if table.contains_key("workspace") {
+                    table["workspace"]["dependencies"]["dojo"].as_table_mut().unwrap()
+                } else {
+                    table["dependencies"]["dojo"].as_table_mut().unwrap()
+                };
```
[the remark this answers is not in the document]
Comment (glihm) [reply crates/dojo-test-utils/src/compiler.rs:209]: This one is for testing, the if is only here to differentiate workspace than other manifests. I guess it's fine to have it right?

File: crates/dojo-lang/src/utils.rs [hunk 4]
```diff
@@ -36,10 +38,14 @@ pub fn get_package_id(db: &dyn SyntaxGroup) -> Option<String> {
             Err(_) => return Option::None,
         };
 
-        if config.contains_key("package")
-            && config["package"].as_table().unwrap().contains_key("name")
+        if config.contains_key("tool")
+            && config["tool"].as_table().unwrap().contains_key("dojo")
+            && config["tool"]["dojo"].as_table().unwrap().contains_key("world")
+            && config["tool"]["dojo"]["world"].as_table().unwrap().contains_key("namespace")
```
[the remark this answers is not in the document]
Comment (glihm) [reply crates/dojo-lang/src/utils.rs:44]: Fixed in 38e4474fc391d10a7e31cbfb2a837d9c6859cae7.

File: examples/spawn-and-move/src/actions.cairo (hunk 2 above)
[the remark this answers is not in the document]
Comment (glihm) [reply examples/spawn-and-move/src/actions.cairo:39]: > incase we are importing from a library we don't need to specify its namespace?

 Just updated the description of the PR with an example. 👍

File: crates/dojo-lang/src/plugin_test_data/model [hunk 5]
```diff
@@ -313,4917 +313,113 @@ struct ModelWithTupleNoPrimitives {
...
-^********************^
-
-error: Unsupported attribute.
- --> test_src/lib.cairo:44:1
-#[dojo::model(version: 2)]
-^************************^
-
-error: Unsupported attribute.
- --> test_src/lib.cairo:51:1
+//! > expected_diagnostics
```
[the remark this answers is not in the document]
Comment (glihm) [reply crates/dojo-lang/src/plugin_test_data/model]: Good Catch, done in latest changes. 👍

**Comment by glihm** [comment]
The mappings are great, but we're still having an issue.

Usually if a namespace is given in the attribute, this is because we want to override the default namespace...
Hence, in the proposed behavior, the namespace is no longer overriding the default one, even in a crate that is not a library.

So we're still need to find a better solution.

We could also add namespace mapping for models inside our crate into the `Scarb.toml`, and we no longer support `#[dojo::model(namespace: "overriding_nm")]`. This can solves both problem, but a bit less maintainable for big projects with a lot of models.

Would be happy to have a discussion on that with you @remybar @lambda-0x or any contributor around.
